_source/_posts/2020-09-28-adapt-dotnet-app-for-samesite-fix.md

```diff
@@ -0,0 +1,352 @@
+9. APP asks identity info to OKTA
+10. Go to step 4
+
+Will USER be happy to repeat the login every 5 minutes? Probably not.
+
+That's why in step 6 OKTA adds a refresh-token cookie to its answer. So how does this make the flow different?
+
+1. USER browses into a protected page in APP
+2. APP receives this first request but it doesn't know USER yet
+3. APP asks identity info to OKTA
+4. OKTA displays the login page
+5. USER enters her credentials and logs in
+6. OKTA verifies the credentials and gives back to APP an identity token with an expiration time 5 minutes **AND a refresh-token** cookie
+7. APP is happy and returns to the browser the desired page
+8. After 5 minutes the identity token expires
+9. APP asks identity info to OKTA
+10. **OKTA receives the refresh-token cookie, and after verifying it, responds directly a new identity token, without login**
+11. Go to step 7
+
+### The Same Site Policy
+
+One important feature of cookies is that they are domain-aware. What this means is that the browser adds them to a request only when that request is bound to the same domain which initially sent the cookie back. In other words, the refresh-token will be sent from the browser only in requests to okta.com, not other domains.
+
+In step 3, USER is happily surfing APP (.app.com) and a request is sent to Okta (.okta.com). The refresh-token hasn't been created yet, so USER is presented with the Okta login page. In step 6 a refresh-token is returned from Okta (.okta.com), so the token is associated with the domain okta.com.
+
+From now up to step 9, all requests are sent to app.com, so the refresh-token is not sent. But in step 9 the request goes to okta.com, so the refresh-token is sent through. Notice, though, that the browser is still showing a page served by APP, from the app.com domain.  This is what is normally called a *cross-site-request* (CSR).
+
+## Cross-Site Request and the Forgery Danger
+
+CSRs are an important feature that makes users experience better, but are open to misuse in a kind of cyber attack called *cross-site-request-forgery* ([CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery)).
+
+Malicious websites could, in fact, issue a request to a third-party website. As an example:
+
+- You browse www.foo.com, which uses a back-end service at api.foo.com. Some cookies get stored locally, associated with the foo.com domain
+- Then you move to www.baz.com, which issues a [AJAX](https://en.wikipedia.org/wiki/Ajax_(programming))/[Fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) (no change of page) request to api.foo.com. The browser takes all the cookies associated to foo.com and sends them through. If api.foo.com relies only on cookies to determine the legitimacy of the request, it will answer as though the request came from www.foo.com, and this is clearly a security breach. Note that the user in front of the computer wouldn't be aware of this, since it occurs behind the scenes.
+
```

Comment:
nit: lots of domain names that _could_ be wrapped in backticks
(but that _might_ make this section too code block heavy)